File: actito_flutter/ios/actito_flutter/Sources/actito_flutter/Events/ActitoEventManager.swift
//
//  ActitoEventManager.swift
//  actito
//
//  Created by Helder Pinhal on 03/12/2020.
//

import Foundation
import Flutter

class ActitoEventManager {

    static let shared = ActitoEventManager()

    private var channels: [ActitoEventType: FlutterEventChannel] = [:]
    private var streams: [ActitoEventType: ActitoEventStream]

    private init() {
        var streams: [ActitoEventType: ActitoEventStream] = [:]

        ActitoEventType.allCases.forEach { type in
            streams[type] = ActitoEventStream(eventType: type)
        }

        self.streams = streams
    }

    func register(for registrar: FlutterPluginRegistrar) {
        streams.values.forEach { stream in
            if let channel = channels[stream.eventType] {
                channel.setStreamHandler(stream)
            } else {
                let channel = FlutterEventChannel(name: stream.name, binaryMessenger: registrar.messenger(), codec: FlutterJSONMethodCodec.sharedInstance())
                channel.setStreamHandler(stream)

                channels[stream.eventType] = channel
            }
        }
    }

    func unregister(for registrar: FlutterPluginRegistrar) {
        channels.values.forEach { channel in
            channel.setStreamHandler(nil)
        }
    }

    func send(_ event: ActitoEvent) {
        DispatchQueue.main.async { [weak self] in
            self?.streams[event.type]?.send(event)
        }
    }
}

File: actito_flutter/android/src/main/kotlin/com/actito/flutter/events/ActitoEventManager.kt
package com.actito.flutter.events

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.JSONMethodCodec

internal object ActitoEventManager {

    private val streams: Map<ActitoEvent.Type, ActitoEventStream> by lazy {
        ActitoEvent.Type.values().associate {
            it to ActitoEventStream(it)
        }
    }

    fun register(messenger: BinaryMessenger) {
        streams.values.forEach {
            val channel = EventChannel(messenger, it.name, JSONMethodCodec.INSTANCE)
            channel.setStreamHandler(it)
        }
    }

    fun send(event: ActitoEvent) {
        Handler(Looper.getMainLooper()).post {
            streams[event.type]?.emit(event)
        }
    }
}
